README.md
macOS on DELL M3800 and XPS15 (9530)
====================================


This project targets at giving the relatively complete functional macOS for both Dell Precision M3800 and XPS15 9530. Before you start, there's a brief introduction of how to finish powering up macOS on your laptop:

1. Create a vanilla installation disk(removable disk).
2. Install Clover with UEFI only and UEFI64Drivers to the installation disk just created. 
3. Replace the origin Clover folder with the one under my Git/M3800/CLOVER.
4. Patch bios to unlock bios menu.
5. BIOS settings:
```sh
Advanced:
    CPU Configuration/CFG Lock = Disabled
    CPU Configuration/LakeTiny Feature = Enabled

    SATA Operation = AHCI 

Chipset:
    PCH-IO Configuration/XHCI Mode = Smart Auto

    System Agent (SA) Configuration/Graphics Configuration:
        Aperture Size = 512MB
        DVMT Pre-Allocated = 160MB
        DVMT Total Gfx Mem = MAX
NOTE: Once you modify your settings in BIOS(especially Graphics Configuration in SA), you have to remove previous ACPI tables first, redump ACPI tables by press Fn+F4/F4 under Clover, and run deploy.sh again to patch your ACPI tables again.
```
6. Install macOS.
7. Once you finish installation of OS X, you can do the following steps to finish the post installation of OS X.

How to use deploy.sh?
----------------

Download the latest version installation package/directory by entering the following command in a terminal window:

```sh
git clone https://github.com/syscl/M3800
```
This will download the whole installation directory to your current directory(./) and the next step is to change the permissions of the file (add +x) so that it can be run.


```sh
cd M3800
chmod +x ./deploy.sh
```


Run the script in a terminal windows by(Note: You should dump the ACPI tables by pressing F4/Fn+F4 under Clover first and then execute the following command lines):

```sh
./deploy.sh
```

Reboot your OS X to see the change. If you have any problem about the script, try to run deploy in DEBUG mode by 
```sh
./deploy.sh -d
```

Note:
- ```VoodooPS2Controller``` is recommended on macOS ```Sierra```(10.```12+```)
- ```AppleSmartTouchPad``` users(recommend on 10.```11-```): for two finger scrolling you need to change the speed of the Scrolling once to get it work and also have to enable them in Trackpad preferences. 

#### Changelog 
[Change logs](https://github.com/syscl/M3800/blob/M3800/Changelog.md) for detail improvements


Changelog.md
M3800-macOS Changelog
=======================
2020-06-27

- Changed drivers64UEFI to drivers/UEFI

2020-06-19

- Fixed lid wake from 10.13.x-10.15.x
- Updated to Clover 5119
- Switch to a much better VoodooPS2 + VoodooInput
- Removed unused kexts

2019-01-02

- Resolved the interference of Lilu that will force inject external GPU properties that cause the lid wake issue
- Attempt to clean up dead code and configuration to fix issue #109 
- Clean up the boot arguements

2019-01-01

- Fix up the ACPI interpreter error (aka Store (TCAP, \_PR.CPU0._PTC ()) issue)
- Fix up HD4600 (0412) inject issues and Audio (layoutID = 0x1B, 27) inject issue

2018-12-27

- Enable native/true multitouch gestures on PS2 controller credit @kprinssu

2018-10-01

- macOS Mojave ready
- Updated EnableLidwake to 4.0.1
- Fixed HDMI/DP issues (tested on HP 23es) 

2018-07-30

- Updated IntelGraphicsFixup to WhateverGreen

2018-06-10

- Enable lid wake by using [EnableLidwake](https://github.com/syscl/EnableLidWake)

2018-06-02

- Update support for 10.13.5 and 10.13.4
- Update Clover EFI to r4509
- Update apfs.efi to 10.13.4
- Update kext to latest
- Clean up deploy

2018-03-22

- Updated to the latest CPU microcode to address CVE-2017-5715 and associated Intel Reboot issue
- Major kexts clean up; reduce project size

2018-02-17

- Updated ```apfs.efi``` to 10.13.2 with log suppressed
- Manually disable ```AppleRTC``` patch due to Clover will turn it on as default
- Confirmed working flawslessly after updating from ```10.13.0``` to ```10.13.3```, note for working ```clipboard```, you need to re-run deploy to rebuild the ```dyld_shared_cache```. Patch remain the same.

2018-02-12

- Resolved HiDPI scale issue during boot
- Updated Clover to r4411

2018-01-16

- Updated lid wake patch for ```0x0a2e0008``` on ```10.13+```
- Updated VoodooPS2 to fix Trackpad blank issue

2017-10-27

- Added apfs.efi so that users can update from ```HFS+``` to ```APFS``` flawslessly
- Updated Clover to r4259

2017-08-16

- Japanese PS2 support
- fix ALPS glide pint buffering issue
- Resolved Headphone noise after updating to 10.12.4+
- Updated Clover to r4173
- Optimized configuration
- EC0->EC such that AppleBusPowerControllerUSB loaded correctly

2017-06-14

- Sync VoodooPS2 from my [repo](https://github.com/syscl/OS-X-Voodoo-PS2-Controller)
- Mute log spam generated by VoodooPS2

2017-03-03

- Fixed a quirk that cause lid wake malfunction (c) syscl
- Change country code to #a for 867Mbps 5Ghz Wi-Fi
- Fixed up freeze issue (c) vit9696, lvs1974

2017-02-22

- Prevent patching RecoveryHD dmg per call (c) syscl

2017-02-21

- Add port injector for ```MacBookPro11,4(M3800/XPS9530)``` credit syscl 
- Prepare for @icedman 's ```VoodooPS2Controller```(tuned by ```syscl```)
- Update sleep: add deep sleep & deep idle properties as per Pike R. Alpha

2017-01-26

- Enable IOPCIFamily to set tolerance latency for PCI devices (c) syscl

2016-12-8

- Updated HDMI audio setting
- Enable 4ports(0000, 0105, 0204, 0306)
- Sync fixUSB to version 2016-9-17

2016-11-20

- Insert Device (PPMC) will not display on IOReg as real MacBook credit syscl
- Insert Device (PMCR) credit syscl
- Insert Device (SLPB) credit syscl
- Correct internal display connector type from LVDS <02 00 00 00> to eDP <00 04 00 00> credit syscl

2016-10-27

- Fix Intel Haswell HD Video for Safari on Sierra credit vit9696

2016-10-26

- macOS 10.12.1 fixed all keyboard function keys(brightness up/down, volume up/down), please update to the latest version 
- macOS 10.12 fixed the System Preference-->Sound-->Output cause headphone noise issue
- Updated SmartTouchPad config with 3D Touch function by syscl
- Removed useless boot glitches patch on 10.12.x to prevent 3 stage logo change. Actually, Apple seems to fix almost the boot glitch issue for us. Cheers!
- Updated Clover to r3859
- Ready for IOPowerManagement to be released, please wait this week:)

2016-10-10

- Updated Clover to r3793:
- Introduce SMCHelper efi driver
- Fix zero pointer dereferencing

2016-10-4

- Updated AppleHDA patches on 10.12.x credit @vbourachot's resource and syscl's configdat

2016-10-1

- Added UHD patched BIOS for XPS9530 credit @ssmith353 see [here](https://github.com/syscl/M3800/issues/22)

2016-9-15

- Mount points at wake to fix external hard disk mounting issue.
- Updated AppleALC to 1.0.16 with bug fix:
- Fixed a rare kernel panic on initialisation failure
- Fixed a rare lock acquisition issue on 10.12
- Fixed an undefined behaviour when failing to perform an i/o operation
- Guaranteed null termination for readFileToBuffer buffers

2016-7-17

- Added Backup checksum detection to prevent too many backup files
  being created.
- Added pixel unlock for Sierra support.
- Optimised code.

2016-7-15

- Added SMC Version (system) and smc-huronriver for FakeSMC.kext.
- Updated Clover to r3625.

2016-6-24

- Replaced FakeSMC with slice's one for further 10.12 support.
- Updated Clover to r3556 for further 10.12 support.
- Updated AppleALC with my modification and optimisation for M3800/XPS9530.

2016-6-12

- Added i5-4200H model support credit busymilk for his [hint](https://github.com/syscl/M3800/issues/25) reminding me that there's i5 model existed in M3800/XPS9530.

2016-6-10

- New patch for DSDT to fix MEM2 issue and also rename TPMX to MEM2 credit syscl.
- Drop MCFG table and drop DMAR table.
- Auto correct/add SSDT-m-M3800.aml in config.plist. 
- Added UHD patched BIOS for M3800 credit @ssmith353 see [here](https://github.com/syscl/M3800/issues/22).
- Added SSDT-m for injecting ALS0.

2016-5-27

- Fixed HiDPI(QHD+/UHD) Recovery HD by setting BooterConfig = 0x2A(0x00101010).

2016-5-25

- Huge update for the support of Recovery HD for DELL M3800/XPS9530 (c) syscl/lighting/Yating Zhou.
- Reverted back to Clover injection for ig-platform-id(easier to turn off "Inject Intel" under Graphics for update).

2016-5-15

- Fixed sleep watcher dir issue due to permission. 

2016-5-14


Finalstep.sh
#!/bin/sh

#  Finalstep.sh
#  
#
#  Created by lighting/syscl(PCBeta) on 15/8/25.
#

#
# Lead to lid wake
#

export LC_NUMERIC="en_US.UTF-8"

# Repository location
REPO=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )

# Define place
decompile=${REPO}/DSDT/raw/
precompile=${REPO}/DSDT/precompile/
compile=${REPO}/DSDT/compile/
tools=${REPO}/tools/
raw=${REPO}/DSDT/raw
prepare=${REPO}/DSDT/prepare

kextstat |grep -y Azul &> /dev/null && result=0 || result=1
kextstat |grep -y HD5000 &> /dev/null && HD=0 || HD=1

if [[ $result -eq 0 && $HD -eq 0 ]];
    then
        echo "After scrpit finish, reboot system and enjoy your OS X! --syscl PCBeta"
        esp=$(grep "dev" "${REPO}"/efi)
        diskutil mount ${esp}
        plist=/Volumes/EFI/EFI/CLOVER/config.plist
        /usr/libexec/plistbuddy -c "Set ':Graphics:ig-platform-id' 0x0a260006" "${plist}" &> /dev/null
        /usr/libexec/plistbuddy -c "Print"  "${plist}" | grep "ig-platform-id = 0x0a260006" &> /dev/null && changestat=0 || changestat=1
        if [ $changestat == 0 ];then
            sudo touch /System/Library/Extensions && sudo kextcache -u /
            echo "FINISH! REBOOT!"
        else
            echo "Failed, ensure ${esp}/EFI/CLOVER/config.plist has right config"
            echo "start first script again"
        fi
    else
        exit 0
fi

exit 0



status.sh
#!/bin/sh

#
# syscl/Yating Zhou/lighting from bbs.PCBeta.com
# Merge for Dell Precision M3800 and XPS15 (9530).
#
export LC_NUMERIC="en_US.UTF-8"
#
# Display style setting.
#
BOLD="\033[1m"
RED="\033[1;31m"
GREEN="\033[1;32m"
BLUE="\033[1;34m"
OFF="\033[m"

#
# Repository location
#
REPO=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )

#
# Define place
#
decompile="${REPO}/DSDT/raw/"
precompile="${REPO}/DSDT/precompile/"
compile="${REPO}/DSDT/compile/"
tools="${REPO}/tools/"
raw="${REPO}/DSDT/raw"
prepare="${REPO}/DSDT/prepare"

create_dir()
{
    if [ ! -d "$1" ];then
    echo "${BLUE}[Creating directory]${OFF}: $1"
    mkdir "$1"
    fi
}

patch_acpi()
{
#   echo "${BLUE}[$2]${OFF} $3"
    if [ "$2" == "syscl" ]
    then
    "${REPO}"/tools/patchmatic "${REPO}"/DSDT/raw/$1.dsl "${REPO}"/DSDT/patches/$3.txt "${REPO}"/DSDT/raw/$1.dsl
    else
    "${REPO}"/tools/patchmatic "${REPO}"/DSDT/raw/$1.dsl "${REPO}"/DSDT/patches/$2/$3.txt "${REPO}"/DSDT/raw/$1.dsl
    fi
}

status()
{
    $1 >./DSDT/report 2>&1
#
# ------------------------------------------------:----------------------------------------------------------:------------------------------
# `grep -i "Error" ./DSDT/report` == *"0 Errors"* : `grep -i "patch complete" ./DSDT/report` == *"complete"* : ! `test -s ./DSDT/report`
# ------------------------------------------------:----------------------------------------------------------:------------------------------
#         iasl failure                            :             patchmatic failure                           : cp, rm, grep, touch, mk...
# ------------------------------------------------:----------------------------------------------------------:------------------------------
#
    if [[ `grep -i "Error" ./DSDT/report` == *"0 Errors"* || `grep -i "patch complete" ./DSDT/report` == *"complete"* || ! `test -s ./DSDT/report` ]]
    then
    echo "[   ${GREEN}OK${OFF}   ] $2."
    else
    echo "[ ${RED}FAILED${OFF} ] $2."
    grep -i -E "Error    |patchmatic|cp" ./DSDT/report >./DSDT/report.tmp
    cat ./DSDT/report.tmp
    fi
    rm ./DSDT/report.tmp ./DSDT/report &> /dev/null
}

compile_table()
{
    echo "${BLUE}[Compiling]${OFF}: $1.dsl"
    "${REPO}"/tools/iasl -vr -w1 -ve -p "${compile}"$1.aml "${precompile}"$1.dsl
}

#
# Generate define directionaries
#
status "create_dir "${REPO}/DSDT"" "Create ./DSDT"
status "create_dir "${prepare}"" "Create ./DSDT/prepare"
status "create_dir "${precompile}"" "Create ./DSDT/precompile"
status "create_dir "${compile}"" "Create ./DSDT/compile"

#
# Decompile dsdt
#
cd "${REPO}"
echo "[${RED}*${OFF}       ] ${BLUE}Disassembling tables.${OFF}"
status ""${REPO}"/tools/iasl -w1 -da -dl "${REPO}"/DSDT/raw/DSDT.aml "${REPO}"/DSDT/raw/SSDT-*.aml" "Disassemble tables"


# Search specification tables by syscl/Yating Zhou

# DptfTa
for num in $(seq 1 20)
do
grep "DptfTa" "${REPO}"/DSDT/raw/SSDT-${num}.dsl &> /dev/null && result=0 || result=1
if [ "${result}" == 0 ];then
DptfTa=SSDT-$num
fi
done

# SaSSDT
for num in $(seq 1 20)
do
grep "SaSsdt" "${REPO}"/DSDT/raw/SSDT-${num}.dsl &> /dev/null && result=0 || result=1
if [ "${result}" == 0 ];then
SaSsdt=SSDT-$num
fi
done

# SgRef
for num in $(seq 1 20)
do
grep "SgRef" "${REPO}"/DSDT/raw/SSDT-${num}.dsl &> /dev/null && result=0 || result=1
if [ "${result}" == 0 ];then
SgRef=SSDT-$num
fi
done

# OptRef
for num in $(seq 1 20)
do
grep "OptRef" "${REPO}"/DSDT/raw/SSDT-${num}.dsl &> /dev/null && result=0 || result=1
if [ "${result}" == 0 ];then
OptRef=SSDT-$num
fi
done

########################
# DSDT Patches
########################

echo "[${RED}**${OFF}      ] ${BLUE}Patching DSDT.dsl${OFF}"

status "patch_acpi DSDT syntax "fix_PARSEOP_ZERO"" "Fix PARSEOP_ZERO"
status "patch_acpi DSDT syntax "fix_ADBG"" "Fix ADBG Error"
status "patch_acpi DSDT graphics "graphics_Rename-GFX0"" "Rename GFX0 to IGPU"
status "patch_acpi DSDT usb "usb_7-series"" "7-series/8-series USB"
status "patch_acpi DSDT battery "battery_Acer-Aspire-E1-571"" "Acer Aspire E1-571"
status "patch_acpi DSDT system "system_IRQ"" "IRQ Fix"
status "patch_acpi DSDT system "system_SMBUS"" "SMBus Fix"
status "patch_acpi DSDT system "system_OSYS"" "OS Check Fix"
status "patch_acpi DSDT system "system_ADP1"" "AC Adapter Fix"
status "patch_acpi DSDT system "system_MCHC"" "Add MCHC"
status "patch_acpi DSDT system "system_WAK2"" "Fix _WAK Arg0 v2"
status "patch_acpi DSDT system "system_IMEI"" "Add IMEI"
status "patch_acpi DSDT system "system_Mutex"" "Fix Non-zero Mutex"
status "patch_acpi DSDT misc "misc_Haswell-LPC"" "Add Haswell LPC"
status "patch_acpi DSDT syscl "audio_HDEF-layout1"" "Add audio Layout 1"
status "patch_acpi DSDT syscl "audio_B0D3_HDAU"" "Rename B0D3 to HDAU"
status "patch_acpi DSDT syscl "remove_glan"" "Remove GLAN device"

########################
# DptfTa Patches
########################

#echo "${BLUE}[DptfTa]${OFF}: Patching ${DptfTa}.dsl in "${REPO}"/DSDT/raw"
echo "[${RED}***${OFF}     ] ${BLUE}Patching ${DptfTa}.dsl${OFF}"
status "patch_acpi ${DptfTa} syscl "_BST-package-size"" "_BST package size"
status "patch_acpi ${DptfTa} graphics "graphics_Rename-GFX0"" "Rename GFX0 to IGPU"

########################
# SaSsdt Patches
########################

echo "[${RED}****${OFF}    ] ${BLUE}Patching ${SaSsdt}.dsl${OFF}"

status "patch_acpi ${SaSsdt} graphics "graphics_Rename-GFX0"" "Rename GFX0 to IGPU"
status "patch_acpi ${SaSsdt} syscl "syscl_Iris_Pro"" "Rename HD4600 to Iris Pro"
status "patch_acpi ${SaSsdt} graphics "graphics_PNLF_haswell"" "Brightness fix (Haswell)"
status "patch_acpi ${SaSsdt} syscl "audio_B0D3_HDAU"" "Rename B0D3 to HDAU"
status "patch_acpi ${SaSsdt} syscl "audio_Intel_HD4600"" "Insert HDAU device"

########################
# SgRef Patches
########################

echo "[${RED}*****${OFF}   ] ${BLUE}Patching ${SgRef}.dsl${OFF}"
status "patch_acpi ${SgRef} graphics "graphics_Rename-GFX0"" "Rename GFX0 to IGPU"

########################
# OptRef Patches
########################

#echo "${BLUE}[OptRef]${OFF}: Patching SSDT-15 in "${REPO}"/DSDT/raw"
echo "[${RED}******${OFF}  ] ${BLUE}Patching ${OptRef}.dsl${OFF}"
status "patch_acpi ${OptRef} syscl "WMMX-invalid-operands"" "Remove invalid operands"
status "patch_acpi ${OptRef} graphics "graphics_Rename-GFX0"" "Rename GFX0 to IGPU"
status "patch_acpi ${OptRef} syscl "graphics_Disable_Nvidia"" "Disable Nvidia card (Non-operational in OS X)"

########################
# Copying all tables to precompile.
########################

echo "[${RED}*******${OFF} ] ${BLUE}Copying tables to precompile${OFF}."
status "cp "${raw}/"*.dsl "${precompile}"" "Copy tables to precompile"

########################
# Compiling tables
########################

echo "[${RED}********${OFF}] ${BLUE}Compiling tables${OFF}."
status "compile_table "DSDT"" "Compiling DSDT"
status "compile_table "${DptfTa}"" "Compiling DptfTa"
status "compile_table "${SaSsdt}"" "Compiling SaSsdt"
status "compile_table "${SgRef}"" "Compiling SgRef"
status "compile_table "${OptRef}"" "Compiling OptRef"

exit 0